### PR TITLE
Implement jballerina profiler

### LINKED ISSUE #35094: Proposal: JBallerina Runtime Profiler
**Description:**
Runtime does not have built-in capability profile a ballerina program especially during load testing. We can use a java profiler like jprofiler still it is not the best possible option.

One option can be to generate a flame graph of runtime.

### PR DESCRIPTION
**Pull request by KD23243**
## Purpose
Fixes #35094

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [x] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: bvm/ballerina-profiler/build.gradle [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+    manifest {
+        attributes 'Implementation-Title': 'Foobar',
+                'Implementation-Version': '',
+                'Built-By': System.getProperty('user.name'),
+                'Built-Date': new Date(),
+                'Built-JDK': System.getProperty('java.version'),
+                'Main-Class': 'io.ballerina.runtime.profiler.Main'
+    }
+}
```
Comment (warunalakshitha) [inline bvm/ballerina-profiler/build.gradle:58]: Lets add new line

File: bvm/ballerina-profiler/src/main/java/io/ballerina/runtime/profiler/Main.java [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+
+    // Define ANSI escape codes for colored console output
+    public static final String ANSI_RESET = "\u001B[0m";
+    public static final String ANSI_GRAY = "\033[37m";
+    public static final String ANSI_CYAN = "\033[1;38;2;32;182;176m";
+
+    // Define public static variables for the program
+    static long profilerStartTime;
+    static int exitCode = 0;
+    public static final String TEMPJARFILENAME = "temp.jar";
```
Comment (warunalakshitha) [inline bvm/ballerina-profiler/src/main/java/io/ballerina/runtime/profiler/Main.java:66]: Shall we move all constants to seperate Constants class

File: bvm/ballerina-profiler/src/main/java/io/ballerina/runtime/profiler/Main.java [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+        String invalidArgument = "Invalid CLI Argument";
+        if (args.length != 0) {
+            for (int i = 0; i < args.length; i++) {
+                switch (args[i]) {
+                    case "--file":
+                        balJarName = args[i + 1];
+                        if (balJarName.startsWith("[") && balJarName.endsWith("]")) {
+                            balJarName = balJarName.substring(1, balJarName.length() - 1);
+                        } else {
+                            System.out.printf(invalidArgument + "\n");
```
Comment (warunalakshitha) [inline bvm/ballerina-profiler/src/main/java/io/ballerina/runtime/profiler/Main.java:109]: Lets use System.out to OUT sttaic variable

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/RunCommand.java [hunk 4]
```diff
@@ -210,19 +222,73 @@ public void execute() {
...
+                    .executeTasks(project);
+        }
+    }
 
-        taskExecutor.executeTasks(project);
+    private void initiateProfiler(Project project, String[] args) throws java.io.IOException, InterruptedException {
+        String[] profilerCommand;
+        String profilerArguments = String.join(" ", args);
+
+        String profilerSource = System.getenv("BALLERINA_HOME") + "/bre/lib/ballerina-profiler-1.0.jar";
```
Comment (warunalakshitha) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/RunCommand.java:253]: Lets remove hardcoded line separator. We can use Paths.get

File: bvm/ballerina-profiler/src/main/java/io/ballerina/runtime/profiler/codegen/CustomClassVisitor.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+ */
+public class CustomClassVisitor extends ClassVisitor {
+    public CustomClassVisitor(ClassVisitor classVisitor) {
+        super(Opcodes.ASM9, classVisitor);
+    }
+
+    @Override
+    public MethodVisitor visitMethod(int access, String name, String desc, String signature, String[] exceptions) {
+        MethodVisitor methodVisitor = super.visitMethod(access, name, desc, signature, exceptions);
+        if (!name.startsWith("$") && desc.startsWith("(Lio/ballerina/runtime/internal/scheduling/Strand")) {
```
Comment (warunalakshitha) [inline bvm/ballerina-profiler/src/main/java/io/ballerina/runtime/profiler/codegen/CustomClassVisitor.java:40]: Lets move this to Constant

File: tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/profiler/ProfilerTest.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    public void testProfilerExecution() throws BallerinaTestException {
+        String sourceRoot = testFileLocation + "/";
+        String packageName = "testProject";
+        Map<String, String> envProperties = new HashMap<>();
+        bMainInstance.addJavaAgents(envProperties);
+        bMainInstance.runMain("run",
+                new String[]{"--profile", packageName},
+                envProperties,
+                null,
+                null,
```
Comment (HindujaB) [inline tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/profiler/ProfilerTest.java:57]: We can improve the tests further by asserting the log outputs or json file outputs.

File: bvm/ballerina-profiler/src/main/java/io/ballerina/runtime/profiler/runtime/Profiler.java [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+                }
+                result.add(decodeIdentifier(frameString));
+            }
+        }
+
+        return result.toString();
+    }
+
+    // This method takes an encoded identifier string as input and returns the decoded version of it
+    public static String decodeIdentifier(String encodedIdentifier) {
```
[the remark this answers is not in the document]
Comment (KD23243) [reply bvm/ballerina-profiler/src/main/java/io/ballerina/runtime/profiler/runtime/Profiler.java:193]: Unfortunately, the class files in the runtime directory can't directly access the Identifier APIs. During the jar file manipulation, the runtime directory is copied to the jar file and runs separately, resulting in limited access to ballerina internal functions.

**Pull request merged**